### PR TITLE
Big pathfinding update + some const-ing

### PR DESCRIPTION
**Pull request by Coolthulhu**
NPC pathfinding was broken by bashables update (when walls became bashable), because pathfinding relied on obstacles not being bashable at all. This resulted in NPCs pathing through walls/windows and then not bashing them, just standing there watching.

This PR:
- Fixes the above bug
- Allows NPCs to grade "how bashable" things are at pathfinding time (considers NPC's strength and weapon)
- Makes them much less eager to wreck cars
- Allows them to open doors other than the basic `t_door` (including car doors)
- Cleans up the code
- Replaces the vector that functioned as the open set in A\* with a much more proper `priority_queue`.
- Consts some of `map::` methods that could be made const

Issues this PR doesn't fix:
- NPCs blundering into fire 
- Heavily limited area of pathfinding (detours not possible)
- Vehicles not getting pathed through properly

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -1503,7 +1504,9 @@ int map::bash_rating(const int str, const int x, const int y)
         return 10;
     }
 
-    return (10 * (str - bash_min)) / (bash_max - bash_min);
+    int ret = (10 * (str - bash_min)) / (bash_max - bash_min);
+    // Round up to 1, so that desperate NPCs can try to bash down walls
+    return ret > 0 ? ret : 1;
```
Comment (BevapDin) [inline src/map.cpp]: In my opinion `return std::max( ret, 1 );` is much easier to read and understand.

**Comment by Rivet-the-Zombie** [comment]
>NPCs pathing through walls/windows and then not bashing them, just standing there watching

I wondered what was causing that behavior.

It was pretty funny sometimes, though.

**Comment by KA101** [comment]
Haven't seen NPCs open a closed  wooden door with peephole, but they do break windows now.

They also try to shoot through the car door, damaging it and wasting 00 shot, but not doing much v. zeds.  Sorry, Coolthulhu.

**Comment by Coolthulhu** [comment]
Shooting through cars is a totally separate issue, not directly related to pathfinding.

I intended this to be a pathfinding update rather than general NPC AI update (because the latter could grow huuuuuuge).

**Comment by Coolthulhu** [comment]
When making the last commit, I found quite a bit of extra checks in `map` class (for example, `map::bash_rating` checks for furniture and terrain thrice each).

I added a const overload to `veh_at`, because I needed a vehicle check during pathfinding. There are a few functions in `map` that manually do what `veh_at` is supposed to do simply because there was no const overload for it.

I'll clean both of those issues up in some later PR.
For now I'd like to get this one merged. It may be a bit too big to be considered a bugfix, but it does fix a bug nonetheless.

**Comment by kevingranade** [comment]
Spawned a bunch of NPCs and followed them around, they seemed to path through smashable things ok, but also went around when there was an option.  Also saw the trying to shoot through things problem, but I agree that doesn't have anything to do with this PR.

**Comment by KA101** [comment]
Yeah, the other dealbreaker for me was that an NPC who had seen and moved-to-engage a critter on the other side of a door (which then closed, due to the Debug Ghost) wouldn't then open the door to engage the critter, but instead stood there.  *shrug*

**Pull request merged**